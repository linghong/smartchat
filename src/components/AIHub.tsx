import React, { useState } from 'react';
import Link from 'next/link';
import { ChevronDown, ChevronUp, Lightbulb } from 'lucide-react';

interface AIHubLink {
  title: string;
  href: string;
  external?: boolean;
}

const links: AIHubLink[] = [
  { title: 'Model List', href: '#' },
  { title: 'Prompt Templates', href: '#' }
];

const AIHub: React.FC = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div
      className={`w-full ${isOpen ? 'bg-slate-300 ' : 'bg-slate-400'} border-t border-gray-700`}
    >
      <button
        onClick={() => setIsOpen(!isOpen)}
        aria-expanded={isOpen}
        aria-controls="ai-hub-menu"
        className="flex items-center justify-between w-full px-4 py-2 text-left text-gray-600 hover:bg-gray-300 hover:text-gray-600 transition-colors duration-200"
      >
        <span className="flex items-center">
          <Lightbulb size={20} className="mr-2" />
          <span className="font-medium">AI Hub</span>
        </span>
        {isOpen ? (
          <ChevronDown size={20} aria-label="chevron-down" />
        ) : (
          <ChevronUp size={20} aria-label="chevron-up" />
        )}
      </button>
      {isOpen && (
        <div
          role="menu"
          aria-labelledby="ai-hub-button"
          className="bg-slate-500 border-t border-gray-700"
        >
          {links.map((link, index) => (
            <Link
              key={index}
              href={link.href}
              className="flex items-center justify-between px-4 py-2 text-sm text-gray-300 border-t border-gray-400 hover:bg-gray-600 transition-colors duration-200"
              target={link.external ? '_blank' : undefined}
              rel={link.external ? 'noopener noreferrer' : undefined}
              role="menuitem"
            >
              {link.title}
            </Link>
          ))}
        </div>
      )}
    </div>
  );
};

export default AIHub;
