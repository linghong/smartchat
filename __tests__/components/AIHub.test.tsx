import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import AIHub from '@/src/components/AIHub';

// Mock the next/link component
jest.mock('next/link', () => {
  const MockLink = ({ children, ...rest }: any) => {
    return <a {...rest}>{children}</a>;
  };
  MockLink.displayName = 'MockLink';
  return MockLink;
});

describe('AIHub Component', () => {
  it('renders the AIHub button', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    expect(button).toBeInTheDocument();
  });

  it('initially renders with the menu closed', () => {
    render(<AIHub />);
    expect(screen.queryByRole('menu')).not.toBeInTheDocument();
  });

  it('opens the menu when the button is clicked', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    fireEvent.click(button);
    expect(screen.getByRole('menu')).toBeInTheDocument();
  });

  it('closes the menu when the button is clicked again', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    fireEvent.click(button);
    fireEvent.click(button);
    expect(screen.queryByRole('menu')).not.toBeInTheDocument();
  });

  it('renders all links when the menu is open', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    fireEvent.click(button);
    expect(screen.getByText('Model List')).toBeInTheDocument();
    expect(screen.getByText('Prompt Templates')).toBeInTheDocument();
  });

  it('applies the correct classes when the menu is open', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    fireEvent.click(button);
    const container = button.closest('div');
    expect(container).toHaveClass('bg-slate-300');
  });

  it('applies the correct classes when the menu is closed', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    const container = button.closest('div');
    expect(container).toHaveClass('bg-slate-400');
  });

  it('renders the correct icon when the menu is open', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    fireEvent.click(button);
    const icon = button.querySelector('svg[aria-label="chevron-down"]');
    expect(icon).toBeInTheDocument();
  });

  it('renders the correct icon when the menu is closed', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    const icon = button.querySelector('svg[aria-label="chevron-up"]');
    expect(icon).toBeInTheDocument();
  });

  it('sets the correct aria attributes on the button', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    expect(button).toHaveAttribute('aria-expanded', 'false');
    expect(button).toHaveAttribute('aria-controls', 'ai-hub-menu');
  });

  it('updates aria-expanded when the menu is toggled', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    fireEvent.click(button);
    expect(button).toHaveAttribute('aria-expanded', 'true');
    fireEvent.click(button);
    expect(button).toHaveAttribute('aria-expanded', 'false');
  });

  it('renders links with correct attributes', () => {
    render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    fireEvent.click(button);
    const links = screen.getAllByRole('menuitem');
    expect(links).toHaveLength(2);
    links.forEach(link => {
      expect(link).toHaveAttribute('href', '#');
      expect(link).not.toHaveAttribute('target');
      expect(link).not.toHaveAttribute('rel');
    });
  });

  it('matches snapshot when closed', () => {
    const { asFragment } = render(<AIHub />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot when open', () => {
    const { asFragment } = render(<AIHub />);
    const button = screen.getByRole('button', { name: /AI Hub/i });
    fireEvent.click(button);
    expect(asFragment()).toMatchSnapshot();
  });
});
