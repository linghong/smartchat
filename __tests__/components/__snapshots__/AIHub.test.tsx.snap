// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AIHub Component matches snapshot when closed 1`] = `
<DocumentFragment>
  <div
    class="w-full bg-slate-400 border-t border-gray-700"
  >
    <button
      aria-controls="ai-hub-menu"
      aria-expanded="false"
      class="flex items-center justify-between w-full px-4 py-2 text-left text-gray-600 hover:bg-gray-300 hover:text-gray-600 transition-colors duration-200"
    >
      <span
        class="flex items-center"
      >
        <svg
          class="lucide lucide-lightbulb mr-2"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 14c.2-1 .7-1.7 1.5-2.5 1-.9 1.5-2.2 1.5-3.5A6 6 0 0 0 6 8c0 1 .2 2.2 1.5 3.5.7.7 1.3 1.5 1.5 2.5"
          />
          <path
            d="M9 18h6"
          />
          <path
            d="M10 22h4"
          />
        </svg>
        <span
          class="font-medium"
        >
          AI Hub
        </span>
      </span>
      <svg
        aria-label="chevron-up"
        class="lucide lucide-chevron-up"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m18 15-6-6-6 6"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`AIHub Component matches snapshot when open 1`] = `
<DocumentFragment>
  <div
    class="w-full bg-slate-300  border-t border-gray-700"
  >
    <button
      aria-controls="ai-hub-menu"
      aria-expanded="true"
      class="flex items-center justify-between w-full px-4 py-2 text-left text-gray-600 hover:bg-gray-300 hover:text-gray-600 transition-colors duration-200"
    >
      <span
        class="flex items-center"
      >
        <svg
          class="lucide lucide-lightbulb mr-2"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 14c.2-1 .7-1.7 1.5-2.5 1-.9 1.5-2.2 1.5-3.5A6 6 0 0 0 6 8c0 1 .2 2.2 1.5 3.5.7.7 1.3 1.5 1.5 2.5"
          />
          <path
            d="M9 18h6"
          />
          <path
            d="M10 22h4"
          />
        </svg>
        <span
          class="font-medium"
        >
          AI Hub
        </span>
      </span>
      <svg
        aria-label="chevron-down"
        class="lucide lucide-chevron-down"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </button>
    <div
      aria-labelledby="ai-hub-button"
      class="bg-slate-500 border-t border-gray-700"
      role="menu"
    >
      <a
        class="flex items-center justify-between px-4 py-2 text-sm text-gray-300 border-t border-gray-400 hover:bg-gray-600 transition-colors duration-200"
        href="#"
        role="menuitem"
      >
        Model List
      </a>
      <a
        class="flex items-center justify-between px-4 py-2 text-sm text-gray-300 border-t border-gray-400 hover:bg-gray-600 transition-colors duration-200"
        href="#"
        role="menuitem"
      >
        Prompt Templates
      </a>
    </div>
  </div>
</DocumentFragment>
`;
